Add retry for field caps node requests

dnhatn opened the pull request:
This change targets the feature branch: `group-field-caps` (based on 7.x).

This adds a retry mechanism for node-based field caps requests introduced in #77047. Merging index responses on data nodes will be implemented in a follow-up.

dnhatn committed:
Add retry for field caps node requests

elasticmachine commented:
Pinging @elastic/es-search (Team:Search)

ywelsch commented on server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java:
```diff
@@ -293,151 +175,207 @@ private static FieldCapabilitiesRequest prepareRemoteRequest(FieldCapabilitiesRe
+            for (ShardIterator shards : shardsIt) {
+                for (ShardRouting shard : shards) {
+                    final DiscoveryNode node = discoveryNodes.get(shard.currentNodeId());
+                    if (node != null) {
```
this should be always != null when shard.curentNodeId != null, correct?

ywelsch reviewed (commented):
Thanks Nhat. I've done just a quick pass today (didn't get further). I'm wondering if some of the retry logic around shard selection / grouping can be unit-tested (e.g. we currently test that retries ARE happening, but don't test how many etc).

dnhatn committed:
combine dispatch

dnhatn commented:
@ywelsch Thank you for your review. All good points - I am addressing them.

dnhatn committed:
fix compile

dnhatn commented:
@ywelsch I think I have addressed your feedback. I will add some more tests to RequestDispatcher. Would you mind taking another look?

I think we can use the existing `FieldCapabilitiesRequest` instead of introducing `FieldCapabilitiesNodeRequest` when the merging response is implemented. I will consider this in the merge response PR.

dnhatn commented:
run elasticsearch-ci/bwc

dnhatn committed:
Merge remote-tracking branch 'elastic/group-field-caps' into 7x-group-field-caps

dnhatn committed:
Add response builder

jtibshirani reviewed (commented):
Thanks for tackling this @dnhatn. I like that we have a dedicated class now to handle the request dispatching logic. The test coverage also looks great.

The one part I wasn't sure of was the synchronization strategy in `RequestDispatcher`. There is quite a bit of logic guarded under synchronized blocks, especially the one in `execute`. I wonder if it'd be better (and if it's even possible) to rely on atomic integers/ thread-safe collections for this? I haven't identified a concrete concern, just raising it to hear your thoughts.

dnhatn commented:
@ywelsch Please hold off on the review. I am working on merging the responses, and I will integrate it in this PR.

dnhatn commented:
> I wonder if it'd be better (and if it's even possible) to rely on atomic integers/ thread-safe collections for this? I haven't identified a concrete concern, just raising it to hear your thoughts

@jtibshirani Yes, we can go without synchronization.

dnhatn reviewed (commented):
@ywelsch I've updated this PR. The RequestDispatcher and its tests are ready. The merging logic is still WIP. I need to discuss it with you before completing it. Would you please review the RequestDispatcher and the approach of the merging results logic?  I will take a look at your can_match PR tomorrow. Sorry for the delay - I've been focusing on this PR. Thank you!

ywelsch reviewed (commented):
> I am working on merging the responses, and I will integrate it in this PR.

Let's revert that part. It has become too difficult to review this PR, and I think we will need more discussions on the merging logic. Let's not block the node-level action on this, but create a clear list of follow-ups.

dnhatn committed:
Revert "Add response builder"

This reverts commit e15b23eea4bf5ed799208c5dbb0c7e393f93aaaf.

dnhatn committed:
dispatcher only

dnhatn commented:
@ywelsch @jtibshirani Thanks for reviews. This is ready again after I removed the merging logic. Would you mind taking another look?

ywelsch reviewed (approved):
LGTM

jtibshirani reviewed (commented):
This looks good to me too, I just left some small comments.

dnhatn committed:
NoShardAvailableActionException

dnhatn committed:
rename

dnhatn committed:
remove null check

dnhatn committed:
shard action runs on the same threadpool

jtibshirani reviewed (approved):
Looks good to me 🎉

dnhatn committed:
Merge remote-tracking branch 'elastic/group-field-caps' into 7x-group-field-caps

dnhatn commented:
@ywelsch @jtibshirani Thanks so much for your reviews.
